type_inference/type_retriever.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/python
+#
+# Copyright 2023 Logica Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict
+from type_inference.postgresql_type_parser import PostgresTypeToLogicaType
+import psycopg2
+
+
+class TypeRetriever:
+  def __init__(self):
+    self.built_in_types = set()
+    self.name_to_type_cache = dict()
+
+  def InitBuiltInTypes(self, connection_string: str):
+    if self.built_in_types:
+      return
+
+    with psycopg2.connect(connection_string) as conn:
+      with conn.cursor() as cur:
+        # this SQL query returns all primitive types (defined by PostgreSQL directly)
+        cur.execute('''
+  SELECT t.typname as type
+  FROM pg_type t
+      LEFT JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace
+  WHERE n.nspname = 'pg_catalog'
+    AND (
+      t.typrelid = 0 OR (SELECT c.relkind = 'c'
+                         FROM pg_catalog.pg_class c
+                         WHERE c.oid = t.typrelid)
+    )
+    AND NOT EXISTS(
+      SELECT 1
+      FROM pg_catalog.pg_type el
+      WHERE el.oid = t.typelem
+        AND el.typarray = t.oid
+    );''')
+        self.built_in_types.update((t[0] for t in cur.fetchall()))
+
+  def UnpackTypeByCachesOnly(self, udt_type: str) -> str | None:
```

Comment:
I am not sure I fully understand logic, but feels like caching should be accessible simpler.
Usually you just do pattern like:

```
def UnpackTypeWithCache(self, t):
  if t not in self.cache:
    self.cache = self.UnpackType(t)
  return self.cache[t]
```

It should make flow simpler here. Right?

Reply:
Types retrieving "pipeline" looks as follows:

- `UnpackTypes` function is called to get string representation of types
- It looks up the caches (user defined types cache and PostgreSQL-to-Logica mapping) for types which are not presented in any of them
- `PopulateCacheByTypes` function is called to send a batched query for all non cached types and fill the cache
- Then, we form resulting dictionary

So, due to query batching, there is no way for us to do simple caching thing. But I tried to rewrite related logic to be simpler. Feel free to comment on that

Reply:
Oh, I think now I finally understand what you've meant in all that comments about querying and caching. I hope this time the logic is clear. I updated existing query that used to get primitives since it had same types' filtering